pkg/cluster/cluster.go

```diff
@@ -232,6 +238,91 @@ func (c *Cluster) Delete() error {
 	return c.forEachMachine(c.deleteMachine)
 }
 
+// List will generate an output for each machine.
+func (c *Cluster) List(all bool, format string) error {
+	machines := c.gatherMachinesWithFallback(all)
+	formatter, err := getFormatter(format)
+	if err != nil {
+		return err
+	}
+	return formatter.Format(machines)
+}
+
+func (c *Cluster) gatherMachinesWithFallback(all bool) (machines []*Machine) {
+	machines = c.gatherMachinesByAPI(all)
+	// Footloose has no machines running. Falling back to display
+	// cluster related data.
+	if len(machines) < 1 {
+		machines = c.gatherMachinesByCluster()
+	}
+	return
+}
+
+func (c *Cluster) gatherMachinesByAPI(all bool) (machines []*Machine) {
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		log.Error(err.Error())
+		return []*Machine{}
+	}
+
+	args := filters.NewArgs()
+	args.Add("label", "org.weaveworks.owner=footloose")
+	if !all {
+		args.Add("label", "org.weaveworks.cluster="+c.spec.Cluster.Name)
+	}
+	ctx := context.Background()
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{
+		Filters: args,
+	})
+	if err != nil {
+		log.Error(err.Error())
+		return []*Machine{}
+	}
+
+	for _, container := range containers {
+		m := Machine{}
+		spec := config.Machine{}
+		m.name = container.Names[0]
+		inspect, err := cli.ContainerInspect(ctx, container.ID)
+		if err != nil {
+			log.Error(err.Error())
+			return []*Machine{}
+		}
+		m.hostname = inspect.Config.Hostname
+		ports := make(map[int]int)
+		for _, p := range container.Ports {
+			ports[int(p.PrivatePort)] = int(p.PublicPort)
+		}
+		m.ports = ports
+		spec.Cmd = container.Command
+		spec.Image = container.Image
+		var volumes []config.Volume
+		for _, mount := range container.Mounts {
+			v := config.Volume{
+				Type:        string(mount.Type),
+				Source:      mount.Source,
+				Destination: mount.Destination,
+				ReadOnly:    mount.RW,
+			}
+			volumes = append(volumes, v)
+		}
+		spec.Volumes = volumes
+		m.spec = &spec
+		machines = append(machines, &m)
+	}
+	return
+}
+
+func (c *Cluster) gatherMachinesByCluster() (machines []*Machine) {
```

Comment:
I saw that you already have forEachMachine but since I wanted each machine I extracted this. I could use the forEachMachine function but then I would have to resort to a package wide variable to track each machines inspect data that came back. That was my other idea in case I would loose the `--all` option.